### PR TITLE
[Testers Needed] SPU LLVM Improvements

### PR DESCRIPTION
**Pull request by Nekotekina**

### CODE REVIEW

**Comment by AniLeo** [comment]
Red Dead Redemption: Broken graphics and massive performance hit right on the title screen

PR
Screenshot from 2021-09-01 20-10-54

Master
Screenshot from 2021-09-01 20-12-51

**Comment by digitaldude555** [comment]
Screenshot (1360)
Screenshot (1361)
gow 3 demo improved somewhat from limited testing.

**Comment by cipherxof** [comment]
Metal Gear Solid 4's graphics look completely broken. They are relatively fine on master.

Screenshot_20210901_203359

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 1]
```diff
@@ -8275,9 +8321,34 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
...
+
+			// const auto base = bitcast<f64[4]>((bitcast<s64[4]>(b) & mask_bf) | 0x3ff0000000000000ull);
+			// const auto step = fpcast<f64[4]>(bitcast<s64[4]>(b) & mask_sf) * fsplat<f64[4]>(std::exp2(-13.f));
+			// const auto yval = fpcast<f64[4]>(bitcast<s64[4]>(a) & mask_yf) * fsplat<f64[4]>(std::exp2(-19.f));
+			// set_vr(op.rt, bitcast<f64[4]>((bitcast<s64[4]>(b) & mask_se) | (bitcast<s64[4]>(base - step * yval) & ~mask_se)));
+		}
 		else
-			set_vr(op.rt, get_vr<f32[4]>(op.rb));
+		{
+			const auto [a, b] = get_vrs<s32[4]>(op.ra, op.rb);
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:8340]: Thing is with FI instruction is that it "fixes" the result of FRSQEST on realhw. Here FRSQEST does exactly what the combination of FRSQEST+FI does so there is no need in FI's fixup. This patern can be detected and used to eliminate FI in this case.

**Comment by Nekotekina** [comment]
Fixed RDR main menu, needs a retest.

**Comment by AniLeo** [comment]
RDR renders correctly now, starting scene performance is about the same as master on RDR GOTY

PR
Screenshot from 2021-09-02 13-28-54

Master
Screenshot from 2021-09-02 13-30-58

**Comment by MSuih** [comment]
Sourced these images from certain individual who is not able to add comments to this repository

Master:
master_mandelrbot

This PR:
PR_mandelbrot

**Comment by Jonathan44062** [comment]
No difference with Sonic Unleashed 
Master: 
master

PR:
PR neko

**Comment by digitaldude555** [comment]
@DaishinSilviaGT @cipherxof Are gt5 and mgs4 back to normal now?

**Comment by digitaldude555** [comment]
[RPCS3GOW3.zip](https://github.com/RPCS3/rpcs3/files/7117538/RPCS3GOW3.zip)
`F SIG: Thread terminated due to fatal error: Unknown STOP code: 0x10 (op=0x10, Out_MBox=empty)
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\Cell\SPUThread.cpp:4672[:22], in function stop_and_signal) (e=0x00000057[87])` GOW 3 E3 demo stops with this, other games just force close the emulator after latest build.

**Comment by DaishinSilviaGT** [comment]
Now GT5 loads properly
Even gets a pretty good performance boost
PR:
new build
Main:
main build

**Comment by digitaldude555** [comment]
Crashes are gone now.

**Comment by RainbowCookie32** [comment]
TLoU crashes when building SPU cache with

```
E SPU: LLVM: Verification failed at 0xc8f8:
Invalid operands for select instruction!
%600 = select <4 x i1> %566, <4 x i32> %596, <4 x float> %599

F SIG: Thread terminated due to fatal error: Compilation failed
(in file ../rpcs3/Emu/Cell/SPURecompiler.cpp:4975, in function compile)
```

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/7123937/RPCS3.log.gz)
[spu-mega-v1-tane.zip](https://github.com/RPCS3/rpcs3/files/7123944/spu-mega-v1-tane.zip)

**Comment by Nekotekina** [comment]
Thanks, should be fixed now.

**Comment by RainbowCookie32** [comment]
Confirmed fixed, seems to run *a bit* faster than master.

Master:
tlou_master

This PR:
tlou_neko

**Comment by digitaldude555** [comment]
Need windows build.

**Comment by Nekotekina** [comment]
Fixed win build

**Comment by dante3732** [comment]
This build useless in this games
<details>
before
after
before
after
before
after
before
after
</details>

**Comment by AniLeo** [comment]
When testing PRs like this, enable the performance overlay with high detail so we can accurately compare the average, min and max FPS values, and also specify what are your specs and/or include a log file

**Comment by dante3732** [comment]
Windows 10 Pro x64 20H2
AMD Ryzen 5 3600 4.2 Ghz
MSI Radeon VEGA 64 8 Gb HBM2
2x8 Gb DDR4 3200

Снимок экрана (2031)
Снимок экрана (2026)

**Comment by MsDarkLow** [comment]
8700k @5.1GHz - In mandelbrot in the PR I get about 4 less fps than in master. Pretty consistent on this.
PR: 119fps - Master: 123fps
<details>
<summary>Mandelbrot screenshots</summary>

Master - 123 fps avg
master1

PR - 119 fps avg
pr1

</details>

**Comment by DefaltBR** [comment]
5600X @4.65Ghz - I'm getting basically equal performace to master in most tests, sometimes a bit less, sometimes a bit more...

<details>
<summary>The Last of Us</summary>
<br>
PR
TLoU_PR
Master
TLoU_Master
</details>

<details>
<summary>God of War Ascension</summary>
<br>
PR
GoWA_PR
Master
GoWA_Master
</details>

**Comment by DefaltBR** [comment]
5600X @4.65Ghz - I'm getting basically equal performance to master in most tests... Sometimes a bit less, sometimes a bit more fps

TLoU_PR
TLoU_Master
GoWA_PR
GoWA_Master

**Comment by Joaozin-tech** [comment]
Captura de tela 2021-09-11 223110
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7148838/RPCS3.log)
[TTY.log](https://github.com/RPCS3/rpcs3/files/7148839/TTY.log)

**Comment by AniLeo** [comment]
Previous post is unrelated to the PR.

**Comment by Joaozin-tech** [comment]
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7157493/RPCS3.log)
[TTY.log](https://github.com/RPCS3/rpcs3/files/7157494/TTY.log)
The Last Of US SPU mega

**Comment by RainbowCookie32** [comment]
Can't reproduce TLoU's crash here. Runs just like I reported on my previous comment

**Comment by Joaozin-tech** [comment]
good to me it occurred emulator closes alone

**Comment by RainbowCookie32** [comment]
Tested mandelbrot on Windows and Linux on the same system.

<details>
<summary>Windows</summary>
<br>

Master
image

PR
image

</details>

<details>
<summary>Linux</summary>
<br>

Master
master

PR
neko

</details>

Linux on master is a tad faster than Windows, but on the PR it's about 20fps faster.
Windows, on the other hand, is about 5-10fps slower on the PR compared to master.

**Comment by Nekotekina** [comment]
That's pretty strange, can you upload windows SPU LLVM cache? It's created under SPU Debug option.

**Comment by MSuih** [comment]
Aphelion's asleep, but I got these from one helpful CPU enthusiast who cannot send comments to here.

[MASTER_CACHE.zip](https://github.com/RPCS3/rpcs3/files/7159825/MASTER_CACHE.zip)
[PR_CACHE.zip](https://github.com/RPCS3/rpcs3/files/7159827/PR_CACHE.zip)

**Comment by Nekotekina** [comment]
Thanks, also it may be worth trying to test with Null renderer for comparison.

**Comment by RainbowCookie32** [comment]
Tested Linux and Windows with null renderer.

On Linux, it's at 120fps on master, 137fps on this PR.
On Windows, it's at 115-120fps on master, 110-115fps on this PR.

Also attached SPU cache and logs with SPU Debug enabled for both Linux and Windows, master and PR.

[SPU Debug - Master Linux.zip](https://github.com/RPCS3/rpcs3/files/7162920/SPU.Debug.-.Master.Linux.zip)
[SPU Debug - Master Windows.zip](https://github.com/RPCS3/rpcs3/files/7162921/SPU.Debug.-.Master.Windows.zip)

[SPU Debug - PR Linux.zip](https://github.com/RPCS3/rpcs3/files/7162922/SPU.Debug.-.PR.Linux.zip)
[SPU Debug - PR Windows.zip](https://github.com/RPCS3/rpcs3/files/7162923/SPU.Debug.-.PR.Windows.zip)

**Comment by Augusto7743** [comment]
10809 breaking GOWA graphics.
https://user-images.githubusercontent.com/21184987/190843553-c82675ca-0683-4f24-9cdf-4f44f9e90161.jpg
https://user-images.githubusercontent.com/21184987/190843558-2802e258-ca4f-4eb4-98b0-c7115f653aac.jpg

**Comment by Augusto7743** [comment]
Tested in recent rpcs3-v0.0.27-14980
The bug with graphics in PR 12670 related with PR 10809 (SPU LLVM Improvements) was fixed.
Thanks !
